File: backend/services/parser.py
import re
from dataclasses import dataclass
from typing import Optional, List

@dataclass
class ParsedMedia:
    title: str
    year: Optional[int] = None
    season: Optional[int] = None
    episode: Optional[int] = None
    extra_tags: List[str] = None
    is_tv: bool = False

# Regexes based on plan specifications
MOVIE_RE = re.compile(r'^(.*?)(?:[. (\[]+((?:19|20)\d{2}))[. )\]]*(.*)$', re.IGNORECASE)
EPISODE_RE = re.compile(r'^(.*?)[\. _-]*S(\d{2})E(\d{2})[\. _-]?(.*)$', re.IGNORECASE)

def parse_filename(filename: str) -> ParsedMedia:
    """Extracts title, year, season, episode, and tags from media filenames."""
    # Cleanup extension
    name = filename.rsplit('.', 1)[0] if '.' in filename else filename
    
    # Try episode first
    ep_match = EPISODE_RE.match(name)
    if ep_match:
        title, season, episode, tags = ep_match.groups()
        return ParsedMedia(
            title=title.replace('.', ' ').strip(),
            season=int(season),
            episode=int(episode),
            extra_tags=[t for t in re.split(r'[\. \-\[\]()]+', tags) if t],
            is_tv=True
        )
        
    # Try movie
    mov_match = MOVIE_RE.match(name)
    if mov_match:
        title, year, tags = mov_match.groups()
        return ParsedMedia(
            title=title.replace('.', ' ').strip(),
            year=int(year),
            extra_tags=[t for t in re.split(r'[\. \-\[\]()]+', tags) if t],
            is_tv=False
        )
        
    # Fallback to plain name
    return ParsedMedia(title=name.replace('.', ' ').strip())

File: internal/services/parser.go
package services

import (
	"regexp"
	"strconv"
	"strings"
)

type ParsedMedia struct {
	Title     string
	Year      int
	Season    int
	Episode   int
	IsTV      bool
	ExtraTags []string
}

var (
	movieRE   = regexp.MustCompile(`(?i)^(.*?)(?:[. (\[]+((?:19|20)\d{2}))[. )\]]*(.*)$`)
	episodeRE = regexp.MustCompile(`(?i)^(.*?)[\. _-]*S(\d{2})E(\d{2})[\. _-]?(.*)$`)
	tagRE     = regexp.MustCompile(`[\. \-\[\]()]+`)
)

func ParseFilename(filename string) ParsedMedia {
	// Remove extension
	name := filename
	if idx := strings.LastIndex(filename, "."); idx != -1 {
		name = filename[:idx]
	}

	// Try Episode first
	if epMatch := episodeRE.FindStringSubmatch(name); epMatch != nil {
		title := strings.ReplaceAll(epMatch[1], ".", " ")
		season, _ := strconv.Atoi(epMatch[2])
		episode, _ := strconv.Atoi(epMatch[3])
		tags := tagRE.Split(epMatch[4], -1)
		
		var cleanTags []string
		for _, t := range tags {
			if t != "" {
				cleanTags = append(cleanTags, t)
			}
		}

		return ParsedMedia{
			Title:     strings.TrimSpace(title),
			Season:    season,
			Episode:   episode,
			IsTV:      true,
			ExtraTags: cleanTags,
		}
	}

	// Try Movie
	if movMatch := movieRE.FindStringSubmatch(name); movMatch != nil {
		title := strings.ReplaceAll(movMatch[1], ".", " ")
		year, _ := strconv.Atoi(movMatch[2])
		tags := tagRE.Split(movMatch[3], -1)

		var cleanTags []string
		for _, t := range tags {
			if t != "" {
				cleanTags = append(cleanTags, t)
			}
		}

		return ParsedMedia{
			Title:     strings.TrimSpace(title),
			Year:      year,
			IsTV:      false,
			ExtraTags: cleanTags,
		}
	}

	// Fallback
	return ParsedMedia{
		Title: strings.TrimSpace(strings.ReplaceAll(name, ".", " ")),
	}
}
